scratch/evaluate_scheduler_qd_dense_scenario.cc

```diff
@@ -0,0 +1,825 @@
+{
+  NS_LOG_DEBUG ("Ending SP with source=" << srcAddr << ", dest=" << destAddr << ", isSource=" << isSource);
+}
+
+void
+ADDTSResponseReceived (Ptr<Node> node, Mac48Address address, StatusCode status, DmgTspecElement element)
+{
+  NS_LOG_DEBUG ("DMG STA=" << address << " received ADDTS response with status=" << status.IsSuccess ());
+  if (status.IsSuccess () || (schedulerType == "ns3::CbapOnlyDmgWifiScheduler"))
+    {
+      CommunicationPairListI it = communicationPairList.find (node);
+      if (it != communicationPairList.end ())
+        {
+          NS_LOG_DEBUG ("Starting APP at node with Id=" << node->GetId ());
+          it->second.startTime = Simulator::Now ();
+          it->second.srcApp->StartApplication ();
+        }
+      else
+        {
+          NS_FATAL_ERROR ("Could not find application to start.");
+        }
+    }
+}
+
+uint32_t
+ComputeServicePeriodDuration (const uint64_t &dataBitRate, const uint64_t &phyModeDataRate)
+{
+  NS_LOG_FUNCTION (dataBitRate << phyModeDataRate);
+  double numberBIs = Seconds (1).GetMicroSeconds () / double (apWifiMac->GetBeaconInterval ().GetMicroSeconds ());
+  uint32_t spDuration = ceil (dataBitRate / double (phyModeDataRate) / numberBIs * 1e6);
+  return spDuration + 1500;
+}
+
+DmgTspecElement
+GetDmgTspecElement (uint8_t allocId, bool isPseudoStatic, uint32_t minAllocation, uint32_t maxAllocation)
+{
+  /* Simple assert for the moment */
+  NS_LOG_FUNCTION (+allocId << isPseudoStatic << minAllocation << maxAllocation);
+  NS_ASSERT_MSG (minAllocation <= maxAllocation, "Minimum Allocation cannot be greater than Maximum Allocation");
+  NS_ASSERT_MSG (maxAllocation <= MAX_SP_BLOCK_DURATION, "Maximum Allocation exceeds Max SP block duration");
+  DmgTspecElement element;
+  DmgAllocationInfo info;
+  info.SetAllocationID (allocId);
+  info.SetAllocationType (SERVICE_PERIOD_ALLOCATION);
+  info.SetAllocationFormat (ISOCHRONOUS);
+  info.SetAsPseudoStatic (isPseudoStatic);
+  info.SetDestinationAid (AID_AP);
+  element.SetDmgAllocationInfo (info);
+  element.SetMinimumAllocation (minAllocation);
+  element.SetMaximumAllocation (maxAllocation);
+  element.SetMinimumDuration (minAllocation);
+  return element;
+}
+
+void
+StationAssociated (Ptr<Node> node, Ptr<DmgStaWifiMac> staWifiMac, Mac48Address address, uint16_t aid)
+{
+  NS_LOG_FUNCTION (node << staWifiMac << address << aid);
+  if (!csv)
+    {
+      NS_LOG_DEBUG ("DMG STA=" << staWifiMac->GetAddress () << " associated with DMG PCP/AP=" << address
+                    << ", AID=" << aid);
+    }
+  /* Send ADDTS request to the PCP/AP */
+  CommunicationPairListI it = communicationPairList.find (node);
+  if (it != communicationPairList.end ())
+    {
+      uint32_t spDuration = ComputeServicePeriodDuration (it->second.appDataRate, WifiMode (phyMode).GetPhyRate ());
+      staWifiMac->CreateAllocation (GetDmgTspecElement (allocationId++, true, spDuration, spDuration));
```

Comment:
doesn't every STA have its own allocation ID? Why are you incrementing it?

Reply:
No, the allocationID is a parameter associated to the DMG Tspec. Each STA has its own AID.

Reply:
yeah, but i mean: a TS is identified by the tuple (srcAid, dstAid, allocationId), right?
Since you have N STAs with a single TS each, you could keep a constant allocationId.
Think about what would happen if there were more than 16 STAs: allocationId would overflow (4 bits if i remember correctly)

Reply:
Oh sure! That's right